mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1167,6 +1197,223 @@ pub async fn p2sh_spending_tx<T: UtxoCommonOps>(coin: &T, input: P2SHSpendingTxI
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(|e| TxGenError::AddressDerivation(format!("Failed to derive premium_address: {}", e)))?;
+    let premium_output = TransactionOutput {
+        value: premium_sat,
+        script_pubkey: Builder::build_p2pkh(&premium_address.hash).to_bytes(),
+    };
+
+    p2sh_spending_tx_preimage(coin, &prev_tx, lock_time, SEQUENCE_FINAL, vec![
+        dex_fee_output,
+        premium_output,
+    ])
+    .await
+    .map_to_mm(TxGenError::Legacy)
+}
+
+pub async fn gen_and_sign_dex_fee_spend_preimage<T: UtxoCommonOps>(
+    coin: &T,
+    args: &GenDexFeeSpendArgs<'_>,
+    htlc_keypair: &KeyPair,
+) -> GenAndSignDexFeeSpendResult {
+    let maker_pub = Public::from_slice(args.maker_pub).map_to_mm(|e| TxGenError::InvalidPubkey(e.to_string()))?;
+    let taker_pub = Public::from_slice(args.taker_pub).map_to_mm(|e| TxGenError::InvalidPubkey(e.to_string()))?;
+
+    let preimage = gen_dex_fee_spend_preimage(
+        coin,
+        args,
+        LocktimeSetting::CalcByHtlcLocktime(args.time_lock),
+        CalcPremiumBy::DeductMinerFee,
+    )
+    .await?;
+
+    let redeem_script = swap_proto_v2_scripts::dex_fee_script(args.time_lock, args.secret_hash, &taker_pub, &maker_pub);
+    let signature = calc_and_sign_sighash(
+        &preimage,
+        DEFAULT_SWAP_VOUT,
+        &redeem_script,
+        htlc_keypair,
+        coin.as_ref().conf.signature_version,
+        coin.as_ref().conf.fork_id,
+    )?;
+    let preimage_tx: UtxoTx = preimage.into();
+    Ok(TxPreimageWithSig {
+        preimage: serialize(&preimage_tx).take(),
+        signature: signature.take(),
+    })
+}
+
+pub async fn validate_dex_fee_spend_preimage<T: UtxoCommonOps + SwapOps>(
+    coin: &T,
+    gen_args: &GenDexFeeSpendArgs<'_>,
+    preimage: &TxPreimageWithSig,
+) -> ValidateDexFeeSpendPreimageResult {
+    // TODO validate that preimage has exactly 2 outputs
+    let actual_preimage_tx: UtxoTx = deserialize(preimage.preimage.as_slice())
+        .map_to_mm(|e| ValidateDexFeeSpendPreimageError::TxDeserialization(e.to_string()))?;
+
+    let maker_pub = Public::from_slice(gen_args.maker_pub)
+        .map_to_mm(|e| ValidateDexFeeSpendPreimageError::InvalidPubkey(e.to_string()))?;
+    let taker_pub = Public::from_slice(gen_args.taker_pub)
+        .map_to_mm(|e| ValidateDexFeeSpendPreimageError::InvalidPubkey(e.to_string()))?;
+
+    // TODO validate premium amount. Might be a bit tricky in the case of dynamic miner fee
+    // TODO validate that output amounts are larger than dust
+
+    // Here, we have to use the exact lock time and premium amount from the preimage because maker
+    // can get different values (e.g. if MTP advances during preimage exchange/fee rate changes)
+    let expected_preimage = gen_dex_fee_spend_preimage(
+        coin,
+        gen_args,
+        LocktimeSetting::UseExact(actual_preimage_tx.lock_time),
+        CalcPremiumBy::UseExactAmount(actual_preimage_tx.outputs[1].value),
```

Comment:
Could you use `get` to access indexes(including the other ones in this module)? It's panic-safe and allows us to handle `out of bounds` situations.

Reply:
I have fixed this line but preferred to keep other cases as is for now to not make the diff even larger. Added to checklist for next iteration: https://github.com/KomodoPlatform/komodo-defi-framework/issues/1895#issue-1786470565